jupyterhub_chameleon/trovi.py

```diff
@@ -0,0 +1,88 @@
+import logging
+import os
+from urllib.parse import urljoin
+
+import requests
+from tornado.web import HTTPError
+
+LOG = logging.getLogger(__name__)
+
+def exchange_token(access_token):
+    # Exchange the user's keystone token for a Trovi token
+    token_resp = requests.post(
+        urljoin(os.getenv("TROVI_URL"), "/token/"),
+        headers={
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        },
+        json={
+            "grant_type": "token_exchange",
+            "subject_token": access_token,
+            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+            "scope": "artifacts:read artifacts:write",
+        },
+    )
+    LOG.debug(f"Trovi token exchange response: {token_resp}")
+
+    if token_resp.status_code in (
+            requests.codes.unauthorized,
+            requests.codes.forbidden,
+    ):
+        LOG.error(f"Authentication to trovi failed: {token_resp}")
+        raise HTTPError(
```

Comment:
Made a comment on this already, but I think this should probably be more specific exception to Trovi, or some `AuthenticationError` that can be interpreted accordingly in the handlers and translated into HTTP response codes.